sa/sa.go

```diff
@@ -1549,9 +1571,164 @@ func (ssa *SQLStorageAuthority) GetOrder(ctx context.Context, req *sapb.OrderReq
+	}
+
+	// Keep a count of the authorizations seen
+	invalidAuthzs := 0
+	deactivatedAuthzs := 0
+	pendingAuthzs := 0
+	validAuthzs := 0
+
+	// Loop over each of the order's authorization objects to examine the authz status
+	for _, authz := range authzs {
+		switch authz.Status {
+		case core.StatusInvalid:
+			invalidAuthzs++
+		case core.StatusDeactivated:
+			deactivatedAuthzs++
+		case core.StatusPending:
+			pendingAuthzs++
+		case core.StatusValid:
+			validAuthzs++
+		default:
+			return "", berrors.InternalServerError(
+				"Order is in an invalid state. Authz %s has invalid status %q",
+				authz.ID, authz.Status)
+		}
+	}
+
+	// An order is invalid if **any** of its authzs are invalid
+	if invalidAuthzs > 0 {
+		return string(core.StatusInvalid), nil
+	}
+	// An order is deactivated if **any** of its authzs are deactivated
+	if deactivatedAuthzs > 0 {
+		return string(core.StatusDeactivated), nil
+	}
+	// An order is pending if **any** of its authzs are pending
+	if pendingAuthzs > 0 {
+		return string(core.StatusPending), nil
+	}
+
+	// An order is fully authorizsed if it has valid authzs for each of the order
+	// names
+	fullyAuthorized := len(order.Names) == validAuthzs
+
+	// If the order isn't fully authorized we've encountered an internal error:
+	// Above we checked for any invalid or pending authzs and should have returned
+	// early. Somehow we made it this far but also don't have the correct number
+	// of valid authzs.
+	if !fullyAuthorized {
+		return "", berrors.InternalServerError(
+			"Order has the incorrect number of valid authorizations & no pending, " +
+				"deactivated or invalid authorizations")
+	}
+
+	// If the order is fully authorized and the certificate serial is set then the
+	// order is valid
+	if fullyAuthorized && order.CertificateSerial != nil && *order.CertificateSerial != "" {
+		return string(core.StatusValid), nil
+	}
+
+	// If the order is fully authorized, and we have began processing it, then the
+	// order is processing.
+	if fullyAuthorized && order.BeganProcessing != nil && *order.BeganProcessing {
+		return string(core.StatusProcessing), nil
+	}
+
+	// If the order is fully authorized, and we haven't begun processing it, then
+	// the order is still pending waiting a finalization request.
+	if fullyAuthorized && order.BeganProcessing != nil && !*order.BeganProcessing {
+		return string(core.StatusPending), nil
+	}
+
+	return "", berrors.InternalServerError(
+		"Order %d is in an invalid state. No state known for this order's "+
+			"authorizations", *order.Id)
+}
+
+func (ssa *SQLStorageAuthority) getAllOrderAuthorizations(
+	ctx context.Context,
+	orderID, acctID int64) (map[string]*core.Authorization, error) {
+
+	now := ssa.clk.Now()
+	var allAuthzs []*core.Authorization
+
+	for _, table := range authorizationTables {
+		var authzs []*core.Authorization
+		_, err := ssa.dbMap.Select(
+			&authzs,
+			fmt.Sprintf(`SELECT %s from %s AS authz
+		LEFT JOIN orderToAuthz
```

Comment:
I think if we do this with a JOIN, it needs to be an INNER JOIN.

Reply:
Fixed in 8b6dd33